### PR TITLE
syncer: remove syncer dependency from shardddl

### PR DESCRIPTION
**Pull request by GMHDBJD**
### What problem does this PR solve?

Issue Number: ref #4287

### What is changed and how it works?
- combine ddl.go and shardddl.go
- shardddl only manages its own state and does not rely on syncer
### Note: some code will be refactor in later pr
- some callback(handleJobFunc...) will be removed
- some state should only be use in shardddl but not in syncer, like onlineddl... 

#### Tests 

 - Unit test
 - Integration test

### Release note 

```release-note
None
```

### CODE REVIEW

File: dm/syncer/ddl.go [hunk 1]
```diff
@@ -18,18 +18,935 @@ import (
...
+	"go.uber.org/atomic"
+	"go.uber.org/zap"
 )
 
+type ShardDDLStrategy interface {
+	preFilter(ddlInfo *ddlInfo, qec *queryEventContext, sourceTable *filter.Table, targetTable *filter.Table) (bool, error)
+	handleDDL(qec *queryEventContext) error
+}
+
+type ShardDDL struct {
```
Comment (lance6716) [inline dm/syncer/ddl.go:53]: Maybe DDLWorker

File: dm/syncer/ddl.go [hunk 2]
```diff
@@ -18,18 +18,935 @@ import (
...
+	case config.ShardOptimistic:
+		shardDDL.strategy = NewOptimistDDL(&shardDDL.logger, syncer)
+	default:
+		shardDDL.strategy = NewNormalDDL(&shardDDL.logger, syncer)
+	}
+	return shardDDL
+}
+
+type Normal struct {
+	ShardDDLStrategy
```
Comment (lance6716) [inline dm/syncer/ddl.go:118]: why embedded this interface? we can assert Normal/Optimist/Pessimist implement `ShardDDLStrategy`

**Review by lance6716 — COMMENTED** [review]
4 / 10 files viewed

File: dm/syncer/ddl.go [hunk 3]
```diff
@@ -18,18 +18,912 @@ import (
...
+	"github.com/pingcap/tiflow/dm/syncer/metrics"
+	onlineddl "github.com/pingcap/tiflow/dm/syncer/online-ddl-tools"
+	sm "github.com/pingcap/tiflow/dm/syncer/safe-mode"
+	"github.com/pingcap/tiflow/dm/syncer/shardddl"
+	"go.uber.org/atomic"
+	"go.uber.org/zap"
 )
 
+type shardDDLStrategy interface {
+	preFilter(ddlInfo *ddlInfo, qec *queryEventContext, sourceTable *filter.Table, targetTable *filter.Table) (bool, error)
```
Comment (lance6716) [inline dm/syncer/ddl.go:49]: when preFilter returns true, it means we should skip this DDL

File: dm/syncer/ddl.go [hunk 4]
```diff
@@ -18,18 +18,912 @@ import (
...
+				}
+			})
+
+			ddl.logger.Info("execute DDL job",
+				zap.String("event", "query"),
+				zap.Stringer("queryEventContext", qec),
+				zap.String("sourceTableID", sourceTableID),
+				zap.Stringer("operation", shardOp))
+		} else {
+			ddl.logger.Info("ignore DDL job",
```
Comment (lance6716) [inline dm/syncer/ddl.go:662]: seems this is the same with old logic, but why we print "ignore" and continue building the DDL job 😂

**Review by lance6716 — APPROVED** [review]

**Comment by lance6716** [comment]
PTAL @D3Hunter @buchuitoudegou

**Review by D3Hunter — APPROVED** [review]

**Review by buchuitoudegou — APPROVED** [review]

**Pull request merged**